// angular-auth-oidc-client/esm2015/lib/services/oidc.security.config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { of, Subject } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
export class OidcConfigService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
        this._onConfigurationLoaded = new Subject();
    }
    /**
     * @return {?}
     */
    get onConfigurationLoaded() {
        return this._onConfigurationLoaded.asObservable();
    }
    /**
     * @param {?} configUrl
     * @return {?}
     */
    load(configUrl) {
        this.httpClient
            .get(configUrl)
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            this.clientConfiguration = response;
            this.load_using_stsServer(this.clientConfiguration.stsServer);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.error(`OidcConfigService 'load' threw an error on calling ${configUrl}`, error);
            this._onConfigurationLoaded.next(false);
            return of(false);
        })))
            .subscribe();
    }
    /**
     * @param {?} stsServer
     * @return {?}
     */
    load_using_stsServer(stsServer) {
        /** @type {?} */
        const url = `${stsServer}/.well-known/openid-configuration`;
        this.httpClient
            .get(url)
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            this.wellKnownEndpoints = response;
            this._onConfigurationLoaded.next(true);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.error(`OidcConfigService 'load_using_stsServer' threw an error on calling ${stsServer}`, error);
            this._onConfigurationLoaded.next(false);
            return of(false);
        })))
            .subscribe();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    load_using_custom_stsServer(url) {
        this.httpClient
            .get(url)
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            this.wellKnownEndpoints = response;
            this._onConfigurationLoaded.next(true);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.error(`OidcConfigService 'load_using_custom_stsServer' threw an error on calling ${url}`, error);
            this._onConfigurationLoaded.next(false);
            return of(false);
        })))
            .subscribe();
    }
}
OidcConfigService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
OidcConfigService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    OidcConfigService.prototype._onConfigurationLoaded;
    /** @type {?} */
    OidcConfigService.prototype.clientConfiguration;
    /** @type {?} */
    OidcConfigService.prototype.wellKnownEndpoints;
    /**
     * @type {?}
     * @private
     */
    OidcConfigService.prototype.httpClient;
}
