// angular-auth-oidc-client/esm5/lib/services/existing-iframe.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LoggerService } from './oidc.logger.service';
var IFrameService = /** @class */ (function () {
    function IFrameService(loggerService) {
        this.loggerService = loggerService;
    }
    /**
     * @param {?} identifier
     * @return {?}
     */
    IFrameService.prototype.getExistingIFrame = /**
     * @param {?} identifier
     * @return {?}
     */
    function (identifier) {
        /** @type {?} */
        var iFrameOnParent = this.getIFrameFromParentWindow(identifier);
        if (iFrameOnParent) {
            return iFrameOnParent;
        }
        return this.getIFrameFromWindow(identifier);
    };
    /**
     * @param {?} identifier
     * @return {?}
     */
    IFrameService.prototype.addIFrameToWindowBody = /**
     * @param {?} identifier
     * @return {?}
     */
    function (identifier) {
        /** @type {?} */
        var sessionIframe = window.document.createElement('iframe');
        sessionIframe.id = identifier;
        this.loggerService.logDebug(sessionIframe);
        sessionIframe.style.display = 'none';
        window.document.body.appendChild(sessionIframe);
        return sessionIframe;
    };
    /**
     * @private
     * @param {?} identifier
     * @return {?}
     */
    IFrameService.prototype.getIFrameFromParentWindow = /**
     * @private
     * @param {?} identifier
     * @return {?}
     */
    function (identifier) {
        return window.parent.document.getElementById(identifier);
    };
    /**
     * @private
     * @param {?} identifier
     * @return {?}
     */
    IFrameService.prototype.getIFrameFromWindow = /**
     * @private
     * @param {?} identifier
     * @return {?}
     */
    function (identifier) {
        return window.document.getElementById(identifier);
    };
    IFrameService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    IFrameService.ctorParameters = function () { return [
        { type: LoggerService }
    ]; };
    return IFrameService;
}());
export { IFrameService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IFrameService.prototype.loggerService;
}
