// angular-auth-oidc-client/esm5/lib/services/oidc-token-helper.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LoggerService } from './oidc.logger.service';
var TokenHelperService = /** @class */ (function () {
    function TokenHelperService(loggerService) {
        this.loggerService = loggerService;
        this.PARTS_OF_TOKEN = 3;
    }
    /**
     * @param {?} dataIdToken
     * @return {?}
     */
    TokenHelperService.prototype.getTokenExpirationDate = /**
     * @param {?} dataIdToken
     * @return {?}
     */
    function (dataIdToken) {
        if (!dataIdToken.hasOwnProperty('exp')) {
            return new Date();
        }
        /** @type {?} */
        var date = new Date(0);
        date.setUTCSeconds(dataIdToken.exp);
        return date;
    };
    /**
     * @param {?} token
     * @param {?} encoded
     * @return {?}
     */
    TokenHelperService.prototype.getHeaderFromToken = /**
     * @param {?} token
     * @param {?} encoded
     * @return {?}
     */
    function (token, encoded) {
        if (!this.tokenIsValid(token)) {
            return {};
        }
        return this.getPartOfToken(token, 0, encoded);
    };
    /**
     * @param {?} token
     * @param {?} encoded
     * @return {?}
     */
    TokenHelperService.prototype.getPayloadFromToken = /**
     * @param {?} token
     * @param {?} encoded
     * @return {?}
     */
    function (token, encoded) {
        if (!this.tokenIsValid(token)) {
            return {};
        }
        return this.getPartOfToken(token, 1, encoded);
    };
    /**
     * @param {?} token
     * @param {?} encoded
     * @return {?}
     */
    TokenHelperService.prototype.getSignatureFromToken = /**
     * @param {?} token
     * @param {?} encoded
     * @return {?}
     */
    function (token, encoded) {
        if (!this.tokenIsValid(token)) {
            return {};
        }
        return this.getPartOfToken(token, 2, encoded);
    };
    /**
     * @private
     * @param {?} token
     * @param {?} index
     * @param {?} encoded
     * @return {?}
     */
    TokenHelperService.prototype.getPartOfToken = /**
     * @private
     * @param {?} token
     * @param {?} index
     * @param {?} encoded
     * @return {?}
     */
    function (token, index, encoded) {
        /** @type {?} */
        var partOfToken = this.extractPartOfToken(token, index);
        if (encoded) {
            return partOfToken;
        }
        /** @type {?} */
        var result = this.urlBase64Decode(partOfToken);
        return JSON.parse(result);
    };
    /**
     * @private
     * @param {?} str
     * @return {?}
     */
    TokenHelperService.prototype.urlBase64Decode = /**
     * @private
     * @param {?} str
     * @return {?}
     */
    function (str) {
        /** @type {?} */
        var output = str.replace(/-/g, '+').replace(/_/g, '/');
        switch (output.length % 4) {
            case 0:
                break;
            case 2:
                output += '==';
                break;
            case 3:
                output += '=';
                break;
            default:
                throw Error('Illegal base64url string!');
        }
        /** @type {?} */
        var decoded = typeof window !== 'undefined' ? window.atob(output) : new Buffer(output, 'base64').toString('binary');
        try {
            // Going backwards: from bytestream, to percent-encoding, to original string.
            return decodeURIComponent(decoded.split('')
                .map((/**
             * @param {?} c
             * @return {?}
             */
            function (c) { return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2); }))
                .join(''));
        }
        catch (err) {
            return decoded;
        }
    };
    /**
     * @private
     * @param {?} token
     * @return {?}
     */
    TokenHelperService.prototype.tokenIsValid = /**
     * @private
     * @param {?} token
     * @return {?}
     */
    function (token) {
        if (!token) {
            this.loggerService.logError("token '" + token + "' is not valid --> token falsy");
            return false;
        }
        if (!((/** @type {?} */ (token))).includes('.')) {
            this.loggerService.logError("token '" + token + "' is not valid --> no dots included");
            return false;
        }
        /** @type {?} */
        var parts = token.split('.');
        if (parts.length !== this.PARTS_OF_TOKEN) {
            this.loggerService.logError("token '" + token + "' is not valid --> token has t have exact three dots");
            return false;
        }
        return true;
    };
    /**
     * @private
     * @param {?} token
     * @param {?} index
     * @return {?}
     */
    TokenHelperService.prototype.extractPartOfToken = /**
     * @private
     * @param {?} token
     * @param {?} index
     * @return {?}
     */
    function (token, index) {
        return token.split('.')[index];
    };
    TokenHelperService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TokenHelperService.ctorParameters = function () { return [
        { type: LoggerService }
    ]; };
    return TokenHelperService;
}());
export { TokenHelperService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TokenHelperService.prototype.PARTS_OF_TOKEN;
    /**
     * @type {?}
     * @private
     */
    TokenHelperService.prototype.loggerService;
}
