// angular-auth-oidc-client/esm5/lib/services/oidc.security.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { OidcSecurityStorage } from './oidc.security.storage';
var OidcSecurityCommon = /** @class */ (function () {
    function OidcSecurityCommon(oidcSecurityStorage) {
        this.oidcSecurityStorage = oidcSecurityStorage;
        this.storage_auth_result = 'authorizationResult';
        this.storage_access_token = 'authorizationData';
        this.storage_id_token = 'authorizationDataIdToken';
        this.storage_is_authorized = '_isAuthorized';
        this.storage_user_data = 'userData';
        this.storage_auth_nonce = 'authNonce';
        this.storage_code_verifier = 'code_verifier';
        this.storage_auth_state_control = 'authStateControl';
        this.storage_session_state = 'session_state';
        this.storage_silent_renew_running = 'storage_silent_renew_running';
        this.storage_custom_request_params = 'storage_custom_request_params';
    }
    Object.defineProperty(OidcSecurityCommon.prototype, "authResult", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_auth_result);
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_auth_result, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "accessToken", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_access_token) || '';
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_access_token, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "idToken", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_id_token) || '';
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_id_token, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "isAuthorized", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_is_authorized);
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_is_authorized, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "userData", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_user_data);
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_user_data, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "authNonce", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_auth_nonce) || '';
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_auth_nonce, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "code_verifier", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_code_verifier) || '';
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_code_verifier, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "authStateControl", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_auth_state_control) || '';
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_auth_state_control, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "sessionState", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_session_state);
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_session_state, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "silentRenewRunning", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_silent_renew_running) || '';
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_silent_renew_running, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "customRequestParams", {
        get: /**
         * @return {?}
         */
        function () {
            return this.retrieve(this.storage_custom_request_params);
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.store(this.storage_custom_request_params, value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    OidcSecurityCommon.prototype.retrieve = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.oidcSecurityStorage.read(key);
    };
    /**
     * @private
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    OidcSecurityCommon.prototype.store = /**
     * @private
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.oidcSecurityStorage.write(key, value);
    };
    /**
     * @param {?} isRenewProcess
     * @return {?}
     */
    OidcSecurityCommon.prototype.resetStorageData = /**
     * @param {?} isRenewProcess
     * @return {?}
     */
    function (isRenewProcess) {
        if (!isRenewProcess) {
            this.store(this.storage_auth_result, '');
            this.store(this.storage_session_state, '');
            this.store(this.storage_silent_renew_running, '');
            this.store(this.storage_is_authorized, false);
            this.store(this.storage_access_token, '');
            this.store(this.storage_id_token, '');
            this.store(this.storage_user_data, '');
            this.store(this.storage_code_verifier, '');
        }
    };
    /**
     * @return {?}
     */
    OidcSecurityCommon.prototype.getAccessToken = /**
     * @return {?}
     */
    function () {
        return this.retrieve(this.storage_access_token);
    };
    /**
     * @return {?}
     */
    OidcSecurityCommon.prototype.getIdToken = /**
     * @return {?}
     */
    function () {
        return this.retrieve(this.storage_id_token);
    };
    OidcSecurityCommon.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    OidcSecurityCommon.ctorParameters = function () { return [
        { type: OidcSecurityStorage }
    ]; };
    return OidcSecurityCommon;
}());
export { OidcSecurityCommon };
if (false) {
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_auth_result;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_access_token;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_id_token;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_is_authorized;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_user_data;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_auth_nonce;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_code_verifier;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_auth_state_control;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_session_state;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_silent_renew_running;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.storage_custom_request_params;
    /**
     * @type {?}
     * @private
     */
    OidcSecurityCommon.prototype.oidcSecurityStorage;
}
